File: project-euler/455/euler-455-v2.pl
#!/usr/bin/perl

use strict;
use warnings;
use autodie;

use 5.016;

use integer;
use bytes;

my $MOD = 1_000_000_000;

sub exp_mod
{
    my ( $b, $e ) = @_;

    if ( $e == 0 )
    {
        return 1;
    }

    my $rec_p = exp_mod( $b, ( $e >> 1 ) );

    my $ret = $rec_p * $rec_p;

    if ( $e & 0x1 )
    {
        $ret *= $b;
    }

    return ( $ret % $MOD );
}

sub calc_f
{
    my ($n) = @_;

    if ( $n % 10 == 0 )
    {
        return 0;
    }

    my $cache = '';

    my $e     = 0;
    my $power = 1;

    # Find the first cycle len.
    while ( !vec( $cache, $power, 1 ) )
    {
        vec( $cache, $power, 1 ) = 1;
    }
    continue
    {
        $e++;
        ( $power *= $n ) %= $MOD;
    }

    my $found_power = $power;
    my $found_e     = $e;

    $power = 1;
    $e     = 0;
    while ( $power != $found_power )
    {
        $e++;
        ( $power *= $n ) %= $MOD;
    }

    my $cycle_len = $found_e - $e;

    $power = 1;
    my $x = 0;
    for my $e2 ( 0 .. $found_e )
    {
        if ( ( $power - $e2 ) % $cycle_len == 0
            and $power > $x )
        {
            $x = $power;
        }
    }
    continue
    {
        ( $power *= $n ) %= $MOD;
    }

    print "f($n) == $x\n";

    return $x;
}

my $MAX = 1_000_000;
my $sum = 0;

my ( $START, $END ) = @ARGV;

STDOUT->autoflush(1);

foreach my $n ( $START .. $END )
{
    $sum += calc_f($n);
}

say "Sum == $sum";

=head1 COPYRIGHT & LICENSE

Copyright 2018 by Shlomi Fish

This program is distributed under the MIT / Expat License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut

File: project-euler/455/euler-455-v2.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

typedef long long myint_t;

/*
=head1 DESCRIPTION

Let f(n) be the largest positive integer x less than 109 such that the last 9 digits of nx form the number x (including leading zeros), or zero if no such integer exists.

For example:

    f(4) = 411728896 (4411728896 = ...490411728896)
    f(10) = 0
    f(157) = 743757 (157743757 = ...567000743757)
    Σf(n), 2 ≤ n ≤ 103 = 442530011399

Find Σf(n), 2 ≤ n ≤ 106.

=cut

*/

#define MOD 1000000000

static inline myint_t exp_mod(myint_t b, myint_t e)
{
    if (e == 0)
    {
        return 1;
    }

    const myint_t rec_p = exp_mod(b, (e >> 1));

    myint_t ret = rec_p * rec_p;

    if (e & 0x1)
    {
        ret *= b;
    }

    return (ret % MOD);
}

// This assumes an 8-bit byte.
static char cache[(MOD >> 3) + 1];

#define NEXT_POWER() { power = ((power * n) % MOD); }

static inline myint_t calc_f(myint_t n)
{
    myint_t x = 0;
    if (n % 10 != 0)
    {
        memset(cache, '\0', sizeof(cache));
        myint_t e = 0;
        myint_t power = 1;

        // Find the first cycle len.
        while (! ( cache[power>>3] & (1 << (power&(8-1)))))
        {
            cache[power >> 3] |= (1 << (power&(8-1)));
            e++;
            NEXT_POWER();
        }

        const myint_t found_power = power;
        const myint_t found_e = e;

        power = 1;
        e = 0;
        while (power != found_power)
        {
            e++;
            NEXT_POWER();
        }

        myint_t cycle_len = found_e - e;

        power = 1;
        for (myint_t e2=0; e2 <= found_e; e2++)
        {
            if (((power - e2) % cycle_len == 0)
                && ( power > x )
            )
            {
                x = power;
            }
            NEXT_POWER();
        }
    }

    printf("f(%lld) == %lld\n", n, x);
    fflush(stdout);

    return x;
}

int main(int argc, char * argv[])
{
    const myint_t START = atol(argv[1]);
    const myint_t END = atol(argv[2]);

    myint_t sum = 0;
    for (myint_t n = START; n <= END ; n++)
    {
        sum += calc_f(n);
    }
    printf( "Sum == %lld\n", sum);
    fflush(stdout);
    return 0;
}

